### DISCUSSION TITLE
"'furo_pygments is undefined'"

### DISCUSSION THREAD

**Discussion by elliotvaucher** [pradyunsg/furo#293] [Potential Issue] [answered]
Dear @pradyunsg, 

First of all, thanks for this amazing theme. 

My problem is the following : 

- I tried building my docs with your theme's source directly in a directory on my computer : https://www.sphinx-doc.org/en/master/usage/theming.html#:~:text=If%20the%20theme,html_theme%20%3D%20%22blue%22%0Ahtml_theme_path%20%3D%20%5B%22.%22%5D

- When I try to `make html`, I get the following error : 

```
Reason: UndefinedError("'furo_pygments' is undefined")
make: *** [html] Error 2
```

Do you have any idea what could go wrong ? 

P.s : I read all previous issues related to Pygments but none of them seemed to work. 

Thank you very much for your time ! 

Elliot.

**Comment by nikitter** [comment]
Hi there! 
I'm getting the same error:(
Did you find any solution?

**Reply by pradyunsg** [reply]
If you can share the complete Sphinx output as well as a reproducer, that would be great!

**Reply by elliotvaucher** [reply]
@pradyunsg here is the complete output. Not sure what you mean by reproducer, sorry ! 

Thanks again for time and attention. 

make html             
Running Sphinx v4.2.0
loading translations [yes]... not available for built-in messages
WARNING: html_static_path entry '_static' does not exist
myst v0.15.2: MdParserConfig(renderer='sphinx', commonmark_only=False, enable_extensions=['dollarmath'], dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', disable_syntax=[], url_schemes=['http', 'https', 'mailto', 'ftp'], heading_anchors=None, heading_slug_func=None, html_meta=[], footnote_transition=True, substitutions=[], sub_delimiters=['{', '}'], words_per_minute=200)
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 2 source files that are out of date
updating environment: [new config] 2 added, 0 changed, 0 removed
reading sources... [100%] index                                                 
.
. // escaping some warnings due to incorrect file paths in my doc. 
.
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done

Theme error:
An error happened in rendering the page eleve.
Reason: UndefinedError("'furo_pygments' is undefined")
make: *** [html] Error 2

**Reply by nikitter** [reply]
@pradyunsg thanks for your fast response!

Here's my complete output:

make html
Running Sphinx v4.2.0
loading translations [ru]... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 3 source files that are out of date
updating environment: [new config] 3 added, 0 changed, 0 removed
reading sources... [100%] page2                                                 
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [ 33%] index                                                  
**Theme error:**
An error happened in rendering the page index.
Reason: UndefinedError("'furo_pygments' is undefined")
make: *** [html] Error 2

**Reply by nikitter** [reply]
So me and @elliotvaucher are getting the same errors.
I've also tried to reinstall both furo and pygments with `pip uninstall`, but it didn't help.

**Reply by pradyunsg** [reply]
> Not sure what you mean by reproducer, sorry !

That would be step by step instructions that I can follow to reproduce the issue.

See https://myweb.uiowa.edu/pbreheny/reproducible.html for the concepts.

**Reply by elliotvaucher** [reply]
Dear @pradyunsg. Thanks for your time and patience. As far as I'm concerned, I will go with using your theme as a parent from which mine will inherit. I think it's a much cleaner and easier solution. 

For me, you can close this discussion. 

Thanks again for your great work with furo !

**Comment by pradyunsg** [comment]
I still don't have any way to reproduce this error.

If someone could share step-by-step instructions for how to reproduce this, that would be ideal! :)

**Reply by elliotvaucher** [reply]
Here is a reproducer : 

- **Environment** : [pip freeze.txt](https://github.com/pradyunsg/furo/files/7553351/pip.freeze.txt). 
- **Filepath** : 
![2 source](https://user-images.githubusercontent.com/67040832/142174633-d5b8faf9-d06f-46ff-97cb-1b8d84589289.PNG)
![1 root](https://user-images.githubusercontent.com/67040832/142174640-f062c272-fac4-4d29-9680-5b89edfb241b.PNG)
- **Conf.py** : [conf.txt](https://github.com/pradyunsg/furo/files/7553428/conf.txt)
- **Error log** : look at first message of current discussion. 

Hope this helps. In case you need anything else, let me know.

**Reply by grgvn** [reply]
Hi, I am working with @elliotvaucher.

Here's the complete filepath.
[filepath.txt](https://github.com/pradyunsg/furo/files/7553941/filepath.txt)

Thank you so much for the great theme and your answers

**Reply by grgvn** [reply]
Just realized no one gave you the steps to reproduce. We encounter this problem at the very begining of the process. 

We first installed Sphinx (with sphinx-quickstart), the virtualenv and dependencies (see pip freeze.txt).

Then, we just took the Furo theme from the repo, and called it through Conf.py (see file above).

The problem appears when building the HTML with _make html_.

**Reply by kapooramit** [reply]
I am facing the same issue. I copied the theme/furo directory to the docs directory where conf.py is located and then ran make HTML. Same error. Anyway to fix it?

**Reply by pradyunsg** [reply]
Ah! Thanks for providing this context @kapooramit!

Furo can only be used when it is installed directly, since the templates require logic that is a part of the `furo` package. It is not possible to use Furo by copy-pasting the templates into the documentation directory and using them.

**Reply by kapooramit** [reply]
I see. Thanks @pradyunsg .
In that case, how can one edit theme source without injecting custom CSS and JS files? I am interested in modifying the theme source (template, scss, js, extensions). Once done, I would want to create a pip installable package.

**Reply by pradyunsg** [reply]
See https://github.com/pradyunsg/furo/discussions/204.

**Comment by pradyunsg** [comment, marked as answer]
So, based on the discussions here, it seems like this happens when you try to copy Furo's templates into your Sphinx documentation and try to use it from there.

That won't work -- Furo is meant to be used as a Python package, installed within the same environment as Sphinx. Please `pip install furo` to install it.

**Comment by kapooramit** [comment]
But how can I customize it without injecting css and js? In a way, if I fork it, and then I want to customize it, the local environment just doesn't work. I am not able to even build.

- Amit
________________________________
